### PR TITLE
Start to add tests for modules/base/tool

### PR DESCRIPTION
**Pull request by metalmatze**
I started to write some tests for modules/base/tool.
I'm going to add more tomorrow.

### CODE REVIEW

File: modules/base/tool_test.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+func TestEncodeSha1(t *testing.T) {
+	if checksum := EncodeSha1("foobar"); checksum != "8843d7f92416211de9ebb963ff4ce28125932878" {
+		t.Errorf("got the wrong sha1sum for string foobar: %s", checksum)
+	}
+}
+
+func TestShortSha(t *testing.T) {
+	if result := ShortSha("veryverylong"); result != "veryverylo" {
```
Comment (bkcsoft) [inline modules/base/tool_test.go]: please use a hex-value when testing hash-functions 🙂

File: modules/base/tool_test.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		t.Errorf("got the wrong sha1sum for string foobar: %s", result)
```
Comment (bkcsoft) [inline modules/base/tool_test.go]: "wrong short sha1sum"?

File: modules/base/tool_test.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		t.Error("expected GetRandomString to be of len 4")
+	}
+}
+
+// TODO: Test PBKDF2()
+// TODO: Test VerifyTimeLimitCode()
+// TODO: Test CreateTimeLimitCode()
+
+func TestHashEmail(t *testing.T) {
+	if hash := HashEmail("lunny@gitea.io"); hash != "1b6d0c0e124d47ded12cd7115addeb11" {
```
Comment (bkcsoft) [inline modules/base/tool_test.go]: I'd refrain against using real addresses when not necessary, this is why ’example.com’ exists 🙂

**Review by bkcsoft — CHANGES_REQUESTED** [review]
otherwise awesome work 😄

File: modules/base/tool_test.go (hunk 1 above)
Comment (metalmatze) [reply modules/base/tool_test.go]: I thought this was a hash-function too. But if you look closer it only shortens a string. Pretty bad function naming. But I don't want to change the API with the tests. That should be a different PR.

Comment (metalmatze) [reply modules/base/tool_test.go]: https://github.com/go-gitea/gitea/pull/90/commits/7d6e0338085a7c066aa1373536ebd88f1837b839

**Comment by jbub** [comment]
Hey @metalmatze, great work on adding tests, i have noticed that you are in some cases repeating and duplicating assert calls (for example `TestTruncateString`, `TestFileSize` or `TestEllipsisString`), you can remove some of the repetitions using table driven tests http://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go, in go 1.7 it can even speedup the tests by parallelizing them http://golang.rakyll.org/parallelize-test-tables/

File: modules/base/tool_test.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+	setting.EnableFederatedAvatar = true
+	assert.Equal(t,
+		"353cbad9b58e69c96154ad99f92bedc7",
+		AvatarLink("gitea@example.com"),
+	)
+	setting.LibravatarService = libravatar.New()
+	assert.Equal(t,
+		"http://cdn.libravatar.org/avatar/353cbad9b58e69c96154ad99f92bedc7",
+		AvatarLink("gitea@example.com"),
```
Comment (strk) [inline modules/base/tool_test.go:85]: If you need a test for a really federated avatar, try strk@kbt.io

Comment (metalmatze) [reply modules/base/tool_test.go:85]: Doesn't make a difference for the test. right?

**Comment by metalmatze** [comment]
Hey @jbub. Thanks for the hint.
It wasn't much work for me to duplicate those lines. Just a shortcut.
Parallelizing sounds pretty interesting though! :rocket:

**Comment by metalmatze** [comment]
I'm done for now. I think we could merge this and add more tests in another PR. This one is pretty big anyway. Updated all the stuff @bkcsoft said. :+1: 

LGTY? :smile:

File: modules/base/tool_test.go (hunk 4 above)
Comment (strk) [reply modules/base/tool_test.go:85]: well, this one isn't really testing the _federated_ part (I've seen packages just prefixing the hash with "cdn.libravatar.org" and be happy with it, it's not our case, we're actually doing the DNS lookups)

**Comment by strk** [comment]
LGTM

**Comment by andreynering** [comment]
Awesome work, LGTM

File: modules/base/tool_test.go (hunk 1 above)
Comment (bkcsoft) [reply modules/base/tool_test.go]: until it's fixed, please use ShaSums ;)

**Comment by bkcsoft** [comment]
LGTM

**Review by bkcsoft — APPROVED** [review]

**Pull request merged**